commit c846af050f944d584f28bc0de310383003c8096d
Author: Miklos Szeredi <mszeredi@redhat.com>
Date:   Mon Dec 14 15:26:14 2020 +0100

    ovl: check privs before decoding file handle
    
    CAP_DAC_READ_SEARCH is required by open_by_handle_at(2) so check it in
    ovl_decode_real_fh() as well to prevent privilege escalation for
    unprivileged overlay mounts.
    
    [Amir] If the mounter is not capable in init ns, ovl_check_origin() and
    ovl_verify_index() will not function as expected and this will break index
    and nfs export features.  So check capability in ovl_can_decode_fh(), to
    auto disable those features.
    
    Signed-off-by: Miklos Szeredi <mszeredi@redhat.com>

diff --git a/fs/overlayfs/util.c b/fs/overlayfs/util.c
index 44b4b62a8ac8..ced63c79e9dd 100644
--- a/fs/overlayfs/util.c
+++ b/fs/overlayfs/util.c
@@ -50,6 +50,9 @@ const struct cred *ovl_override_creds(struct super_block *sb)
  */
 int ovl_can_decode_fh(struct super_block *sb)
 {
+	if (!capable(CAP_DAC_READ_SEARCH))
+		return 0;
+
 	if (!sb->s_export_op || !sb->s_export_op->fh_to_dentry)
 		return 0;